Unable to use 2FA recovery code

### Prerequisites

- [x] I have searched the existing **Closed _AND_ Open** [Issues](https://github.com/dani-garcia/vaultwarden/issues?q=is%3Aissue%20) **_AND_** [Discussions](https://github.com/dani-garcia/vaultwarden/discussions?discussions_q=)
- [x] I have searched and read the [documentation](https://github.com/dani-garcia/vaultwarden/wiki/)

### Vaultwarden Support String

(admin panel disabled, I can figure this out if it's needed)

### Vaultwarden Build Version

docker.io/vaultwarden/server:1.35.7-alpine

### Deployment method

Official Container Image

### Custom deployment method

_No response_

### Reverse Proxy

ingress-nginx 1.15.1

### Host/Server Operating System

Linux

### Operating System Version

Kubernetes

### Clients

Android, Web Vault, Browser Extension

### Client Version

web 2026.2.0

### Steps To Reproduce

1. Go to /recover-2fa page
2. Enter inputs
3. Hit submit
4. Toasts: 
" Two-step login turned off on your account.
You have been logged in! "
5. 2FA not actually reset, not logged in

### Expected Result

2FA gets reset, I'm able to log in

### Actual Result

400 bad request:

```json
{
  "MasterPasswordPolicy": {
    "Object": "masterPasswordPolicy"
  },
  "TwoFactorProviders": [
    "0"
  ],
  "TwoFactorProviders2": {
    "0": null
  },
  "error": "invalid_grant",
  "error_description": "Two factor required."
}
```

### Logs

```text
[2026-04-14 15:52:29.471][request][INFO] POST /identity/accounts/prelogin
[2026-04-14 15:52:29.474][response][INFO] (prelogin) POST /identity/accounts/prelogin => 200 OK
[2026-04-14 15:52:29.796][request][INFO] POST /identity/connect/token
[2026-04-14 15:52:29.943][error][ERROR] Invalid two factor provider
[2026-04-14 15:52:29.943][response][INFO] (login) POST /identity/connect/token => 400 Bad Request
```

### Screenshots or Videos

_No response_

### Additional Context

My clients got logged out and wanted 2FA validation again, I suspect because of the 2FA troubles in recent releases. However, I use the bitwarden android app for my vaultwarden TOTP (yes, I know). I'd always planned to use the recovery code for a situation like this, but now it's not working.

I can enter the token multiple times on `/#/recover-2fa`. 
The following line is logged when I try:
```
Apr 15 00:44:44 vaultwarden[47849]: [2026-04-15 00:44:44.240][error][ERROR] Invalid two factor provider
```
I'll have a look at increasing log level.

@BlackDex with log level debug:
```
Apr 15 00:50:30 vaultwarden[166448]: [2026-04-15 00:50:30.089][vaultwarden::api::core::two_factor][DEBUG] Sending notifications for incomplete 2FA logins
Apr 15 00:50:30 vaultwarden[166448]: [2026-04-15 00:50:30.089][vaultwarden::api::core::accounts][DEBUG] Purging auth requests
Apr 15 00:51:30 vaultwarden[166448]: [2026-04-15 00:51:30.091][vaultwarden::api::core::accounts][DEBUG] Purging auth requests
Apr 15 00:51:30 vaultwarden[166448]: [2026-04-15 00:51:30.091][vaultwarden::api::core::two_factor][DEBUG] Sending notifications for incomplete 2FA logins
Apr 15 00:51:51 vaultwarden[166448]: [2026-04-15 00:51:51.274][request][INFO] GET /api/config
Apr 15 00:51:51 vaultwarden[166448]: [2026-04-15 00:51:51.275][response][INFO] (config) GET /api/config => 200 OK
Apr 15 00:52:02 vaultwarden[166448]: [2026-04-15 00:52:02.522][request][INFO] POST /identity/accounts/prelogin
Apr 15 00:52:02 vaultwarden[166448]: [2026-04-15 00:52:02.523][response][INFO] (prelogin) POST /identity/accounts/prelogin => 200 OK
Apr 15 00:52:03 vaultwarden[166448]: [2026-04-15 00:52:03.004][request][INFO] POST /identity/connect/token
Apr 15 00:52:03 vaultwarden[166448]: [2026-04-15 00:52:03.081][error][ERROR] Invalid two factor provider
Apr 15 00:52:03 vaultwarden[166448]: [2026-04-15 00:52:03.081][response][INFO] (login) POST /identity/connect/token => 400 Bad Request
```

Response for the post to `/identity/connect/token`:
```
{"MasterPasswordPolicy":{"Object":"masterPasswordPolicy"},"TwoFactorProviders":["0"],"TwoFactorProviders2":{"0":null},"error":"invalid_grant","error_description":"Two factor required."}
```

--- a/src/api/identity.rs
+++ b/src/api/identity.rs
@@ -742,7 +742,7 @@ async fn twofactor_auth(
 
     TwoFactorIncomplete::mark_incomplete(&user.uuid, &device.uuid, &device.name, device.atype, ip, conn).await?;
 
-    let mut twofactor_ids: Vec<_> = twofactors
+    let twofactor_ids: Vec<_> = twofactors
         .iter()
         .filter_map(|tf| {
             let provider_type = TwoFactorType::from_i32(tf.atype)?;
@@ -753,15 +753,11 @@ async fn twofactor_auth(
         err!("No enabled and usable two factor providers are available for this account")
     }
 
-    // Add TwoFactorTypes which are not stored as a record but might be enabled
-    // Since these types could also be not valid, we do some custom checks here
-    twofactor_ids.extend(
-        (!CONFIG.disable_2fa_remember() && device.twofactor_remember.is_some())
-            .then_some(TwoFactorType::Remember as i32),
-    );
-
     let selected_id = data.two_factor_provider.unwrap_or(twofactor_ids[0]); // If we aren't given a two factor provider, assume the first one
-    if !twofactor_ids.contains(&selected_id) {
+                                                                            // Ignore Remember and RecoveryCode Types during this check, these are special
+    if ![TwoFactorType::Remember as i32, TwoFactorType::RecoveryCode as i32].contains(&selected_id)
+        && !twofactor_ids.contains(&selected_id)
+    {
         err_json!(
             _json_err_twofactor(&twofactor_ids, &user.uuid, data, client_version, conn).await?,
             "Invalid two factor provider"

